clients/apply/source_test.go

```diff
@@ -0,0 +1,434 @@
+						encoding: apply.TemplateEncodingYaml,
+					},
+					{
+						name:     filepath.Join(rootDir, "bar", "foo.jsonnet"),
+						contents: "foo.jsonnet",
+						encoding: apply.TemplateEncodingJsonnet,
+					},
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			tmp, err := os.MkdirTemp("", "")
+			require.NoError(t, err)
+			defer os.RemoveAll(tmp)
+			tc.setup(t, tmp)
+
+			sources, err := apply.SourcesFromPath(tc.inPath(tmp), tc.recursive, tc.inEncoding)
+			require.NoError(t, err)
+			expected := tc.expected(tmp)
+			require.Len(t, sources, len(expected))
+
+			sort.Slice(sources, func(i, j int) bool {
+				return sources[i].Name < sources[j].Name
+			})
+			sort.Slice(expected, func(i, j int) bool {
+				return expected[i].name < expected[j].name
+			})
+
+			for i := range expected {
+				source := sources[i]
+				contents := expected[i]
+
+				require.Equal(t, contents.encoding, source.Encoding)
+				sourceIn, err := source.Open(context.Background())
+				require.NoError(t, err)
+				bytes, err := io.ReadAll(sourceIn)
+				sourceIn.Close()
+				require.NoError(t, err)
+				require.Equal(t, contents.contents, string(bytes))
+			}
+		})
+	}
+}
+
+func TestSourceFromURL(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name           string
+		filename       string
+		inEncoding     apply.TemplateEncoding
+		expectEncoding apply.TemplateEncoding
+		resStatus      int
+		resBody        string
+		expectErr      bool
+	}{
+		{
+			name:           "JSON file",
+			filename:       "foo.json",
+			expectEncoding: apply.TemplateEncodingJson,
+			resStatus:      200,
+			resBody:        "Foo bar",
+		},
+		{
+			name:           "YAML file",
+			filename:       "foo.yaml",
+			expectEncoding: apply.TemplateEncodingYaml,
+			resStatus:      200,
+			resBody:        "Foo bar",
+		},
+		{
+			name:           "YML file",
+			filename:       "foo.yml",
+			expectEncoding: apply.TemplateEncodingYaml,
+			resStatus:      200,
+			resBody:        "Foo bar",
+		},
+		{
+			name:           "JSONNET file",
+			filename:       "foo.jsonnet",
+			expectEncoding: apply.TemplateEncodingJsonnet,
+			resStatus:      200,
+			resBody:        "Foo bar",
+		},
+		{
+			name:           "explicit encoding",
+			filename:       "foo",
+			inEncoding:     apply.TemplateEncodingJson,
+			expectEncoding: apply.TemplateEncodingJson,
+			resStatus:      200,
+			resBody:        "Foo bar",
+		},
+		{
+			name:           "err response",
+			filename:       "foo.json",
+			expectEncoding: apply.TemplateEncodingJson,
+			resStatus:      403,
+			resBody:        "OH NO",
+			expectErr:      true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+				rw.WriteHeader(tc.resStatus)
+				rw.Write([]byte(tc.resBody))
+				return
```

Comment:
Looks like the linter is complaining about this; you shouldn't need a return here at the end of a http handler.

Reply:
Ah now it's fixed, nice!